Package and publish groclient library to PyPI with Poetry (CLEWS-15764)

jli opened the pull request:
# Changes

- Backwards incompatible changes:
  - Drops `api.client` package from library.
  - Drops support for Python 2 (no code has changed, the library is just marked as only supporting Python >= 3.6).
- Removes Python 2 from Shippable config (should speed up Shippable builds by 2x).
- Adds Poetry for building the package and publishing to PyPI. Automated via Shippable.
  - New tagged releases cause Shippable to publish a new version to PyPI.
  - PRs merged to development also publish to TestPyPI.

Installing the PyPI package will be much, much faster, since it doesn't require cloning this whole repo (this repo is oddly large and takes a minute to clone - perhaps due to to the jupyter notebooks?).

# Dealing with backwards compatibility issues

If users run into backwards compatibility issues, they can pin to the previous tagged release:
`pip install git+https://github.com/gro-intelligence/api-client.git@v1.82.0`

Or a more recent commit, from before this PR is merged:
`pip install git+https://github.com/gro-intelligence/api-client.git@0055cc2883956223c1885230fbb0f6a165854b7e`


# Testing

All testing was manual.

- Tested Shippable config for building and publish to TestPyPI works.
- Tested installing published TestPyPI package.
- Verified that Python 2 clients will get errors when installing the latest git version after this is submitted, and that this can be worked around by installing previous library version (via version tag or commit hash).

**Not tested**:

- Shippable should build and publish a new release to PyPI once we make a new release on Github.


# Documentation

- Added details to CONTRIBUTING.md with how to use Poetry.
- Updated README and developer docs to install from PyPI instead of via git.

Commit:
update CONTRIBUTING.md for poetry

Commit:
initial pyproject.toml poetry config

Commit:
add LICENSE (MIT)

Commit:
build/publish PyPI package via shippable

drops support for building/testing python 2.7 in shippable.

Commit:
add poetry-dynamic-versioning

sets the version on packages based on git tags

Commit:
add poetry.lock to fix shippable

For some reason, shippable fails on `poetry install` when it needs to do
its own dependency resolution. Note that checking in poetry.lock doesn't
affect users of the client library, it's just used by poetry to install
the right packages locally.

Commit:
update docs/README.md for poetry

Commit:
docs: prefer installing PyPI package instead of from git

Commit:
fix docs formatting typo

Commit:
include deprecated `api` module in package

muzigao commented on README.md:
```diff
@@ -13,11 +13,13 @@ Client library for accessing Gro Intelligence's agricultural data platform.
 Note that even if you are using [Anaconda](https://www.anaconda.com/), the API Client install should still be performed using pip and not [conda](https://docs.conda.io/en/latest/).
 
+If you're unable to access PyPI, you can install the latest code from Github: `pip install git+https://github.com/gro-intelligence/api-client.git`
+
```
do you think we should mention python2 user here and/or installation.rst?

jli replied:
What do you think we should mention? Like, how to install the last Python 2 version?

Perhaps we can wait to see if there are any Python 2 users left before adding documentation for them? WDYT?

muzigao replied:
that should also work

muzigao commented on shippable.yml:
```diff
@@ -1,54 +1,65 @@
+    - >
+      if [ "$BRANCH" == "development" ]; then
+        poetry config repositories.testpypi https://test.pypi.org/legacy/ &&
+        poetry publish -u __token__ -p $TESTPYPI_TOKEN -r testpypi
```
it seems shippable is still failing here

muzigao replied:
what's $TESTPYPI_TOKEN? should it be GROAPI_TOKEN?

jli replied:
**tokens**

`$TESTPYPI_TOKEN` and `$PYPI_TOKEN` are secrets for publishing to PyPI. They are encoded in the encrypted environment variable also added to our shippable.yml file.

**shippable failure**

The shippable failure is confusing. You can see that shippable actually [succeeds for this branch](https://app.shippable.com/github/gro-intelligence/api-client/runs?branchName=jli-poetry2).

But for some reason, shippable also tries to build in the "context" of the development branch. So the failure is due to [job 3945]( https://app.shippable.com/github/gro-intelligence/api-client/runs/3947/1/console) because that job (for some reason) doesn't include the new environment variables I added.

I tried contacting shippable support, and also filed a [bug report](https://github.com/Shippable/support/issues/5165), but haven't heard back yet.

I *think* that if we merge this PR, that should fix the environment variable issue. If it doesn't, we can remove the shippable config for automatically publishing to PyPI and just do that step manually.

What do you think?

Commit:
workaround: disable PyPI publishing in shippable

context: https://github.com/gro-intelligence/api-client/pull/277#discussion_r516229938

my guess is that this commit should fix the shippable build for this PR,
and that we can reenable the poetry pypi publishing in a subsequent PR
because the environment variables will be defined correctly...

muzigao reviewed (approved):
LGTM

muzigao committed:
Merge branch 'development' into jli-poetry2
